infer/apply_defaults.go

```diff
@@ -0,0 +1,354 @@
+// Copyright 2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infer
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+
+	"github.com/pulumi/pulumi-go-provider/internal/introspect"
+)
+
+// The object that controls default application.
+type defaultsWalker struct {
+	// seen is the stack of types that defaultsWalker has descended into.
+	seen []reflect.Type
```

Comment:
This is expressed on the `reflect.*` domain because annotations are defined on go types, not on property values. As an aside, this makes it relatively to handle recursive types, since go types already define the properties we want to bound recursion.

Reply:
If we switched to applying defaults to property values, we would still need the recursive traversal to gather the default values, but would then need to draw another mechanism in to apply the values.

Reply:
This behavior is under test because of this field:

https://github.com/pulumi/pulumi-go-provider/blob/9da507871298b1d0cf80626437761a0f3b32286a/infer/tests/provider.go#L201

Reply:
>  would then need to draw another mechanism in to apply the values.

It would seem that if you're generating default values into Pulumi schema there's already a mechanism somewhere or that, but ymmv, I'm not familiar with the codebase so taking your word for it.